smaug/__init__.py

```diff
@@ -0,0 +1,10 @@
+from smaug.core.types_pb2 import *
+
+from smaug.python.ops import math_ops as math
```

Comment:
Thanks for doing this. It definitely looks a lot better now.

Is there a strategy to how you decided whether to import by module or just import everything? As in, why do we have to write `sg.nn.convolution`, but not `sg.array.flatten`? Not saying the latter is better, but I'd like to have consistency in this hierarchy. One reason I can imagine is that a convolution doesn't have to necessarily be the 4D type used in most networks, so putting that 4d convolution until the `nn` namespace could make sense (we could add a different convolution under a different namespace in the future). Whatever we decide, we should document it here so future us knows what convention to follow.

In general, I like that `Graph` and `Tensor` are top level objects since they are pretty fundamental to SMAUG. I also am fine with promoting one-off functions or otherwise redundant-looking statements to top level like `sg.data.input_data` -> `sg.input_data`. But if a function logically belongs to a group of operations where the number of operations in that group could grow to more than 2 or 3, it might be better off in a second-level namespace, as long as that second-level namespace is short. 

Overall, remember that the user can always import these in their own modules and rename them however they like; our goal is to strike a balance between clarity and succinctness so users don't go off naming everything differently in their code (and then confusing other users with their different API names).

Here's my proposal - feel free to give feedback:

* Top-level: Graph, Tensor, `cond`, `input_data`.
* `sg.nn` for all operators under `nn_ops.py`, `activation.py, `recurrent.py`, `attention.py`.
* `sg.tensor` for all operators under `array_ops.py (e.g. `sg.tensor.reshape`, `sg.tensor.reorder`, etc)
* `sg.math` for all operators under `math_ops.py`. This one is a bit ambiguous since everything is technically a math operation, but most of the stuff here is for elementwise computations. I don't have a better idea though.

Reply:
I think `nn` and `math` are the two groups that would potentially contain many operators (yes we can add different convolutions and other arithmetic operations), so it makes sense to have specific namespaces for them. Except these two groups, the convention for now is to put every other operator under `sg`, maybe in the future we would need a namespace called `sim` for simulation specific stuff. 

Therefore, regarding the proposal, I think the third group could be put to the top level as well? I checked with [TensorFlow](https://www.tensorflow.org/api_docs/python/tf/reshape) and [PyTorch](https://pytorch.org/docs/stable/generated/torch.reshape.html#torch.reshape), they seem to put the `array_ops` under the top level, I guess we should follow this convention.

Reply:
There's seven different tensor reshaping operators in `array_ops`. I think that's good justification for putting them into a separate namespace. We don't need to mirror TF/PyT all the time.

Reply:
Okay, added a namespace `tensor` for `array_ops`.

Reply:
Also all the models have been updated to use the new APIs.

Reply:
We'll have to update the tests too, but feel free to do that in a separate commit.